File: test/language/offsets/python/AutoArrayOffsetTest.py
import zserio

import Offsets

class AutoArrayOffsetTest(Offsets.TestCase):
    def testBitSizeOf(self):
        createWrongOffset = False
        autoArrayHolder = self._createAutoArrayHolder(createWrongOffset)
        self.assertEqual(self.AUTO_ARRAY_HOLDER_BIT_SIZE, autoArrayHolder.bitsizeof())

    def testBitSizeOfWithPosition(self):
        createWrongOffset = False
        autoArrayHolder = self._createAutoArrayHolder(createWrongOffset)
        bitPosition = 2
        self.assertEqual(self.AUTO_ARRAY_HOLDER_BIT_SIZE - bitPosition, autoArrayHolder.bitsizeof(bitPosition))

    def testInitializeOffsets(self):
        createWrongOffset = True
        autoArrayHolder = self._createAutoArrayHolder(createWrongOffset)
        bitPosition = 0
        self.assertEqual(self.AUTO_ARRAY_HOLDER_BIT_SIZE, autoArrayHolder.initialize_offsets(bitPosition))
        self._checkAutoArrayHolder(autoArrayHolder)

    def testInitializeOffsetsWithPosition(self):
        createWrongOffset = True
        autoArrayHolder = self._createAutoArrayHolder(createWrongOffset)
        bitPosition = 2
        self.assertEqual(self.AUTO_ARRAY_HOLDER_BIT_SIZE, autoArrayHolder.initialize_offsets(bitPosition))
        self._checkAutoArrayHolder(autoArrayHolder, bitPosition)

    def testRead(self):
        writeWrongOffset = False
        writer = zserio.BitStreamWriter()
        self._writeAutoArrayHolderToStream(writer, writeWrongOffset)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        autoArrayHolder = self.api.AutoArrayHolder.from_reader(reader)
        self._checkAutoArrayHolder(autoArrayHolder)

    def testReadWrongOffsets(self):
        writeWrongOffset = True
        writer = zserio.BitStreamWriter()
        self._writeAutoArrayHolderToStream(writer, writeWrongOffset)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        with self.assertRaises(zserio.PythonRuntimeException):
            autoArrayHolder = self.api.AutoArrayHolder.from_reader(reader)
            self._checkAutoArrayHolder(autoArrayHolder)

    def testWrite(self):
        createWrongOffset = True
        autoArrayHolder = self._createAutoArrayHolder(createWrongOffset)
        writer = zserio.BitStreamWriter()
        autoArrayHolder.initialize_offsets(writer.bitposition)
        autoArrayHolder.write(writer)
        self._checkAutoArrayHolder(autoArrayHolder)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readAutoArrayHolder = self.api.AutoArrayHolder.from_reader(reader)
        self._checkAutoArrayHolder(readAutoArrayHolder)
        self.assertTrue(autoArrayHolder == readAutoArrayHolder)

    def testWriteWithPosition(self):
        createWrongOffset = True
        autoArrayHolder = self._createAutoArrayHolder(createWrongOffset)
        writer = zserio.BitStreamWriter()
        bitPosition = 2
        writer.write_bits(0, bitPosition)
        autoArrayHolder.initialize_offsets(writer.bitposition)
        autoArrayHolder.write(writer)
        self._checkAutoArrayHolder(autoArrayHolder, bitPosition)

    def testWriteWrongOffset(self):
        createWrongOffset = True
        autoArrayHolder = self._createAutoArrayHolder(createWrongOffset)
        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            autoArrayHolder.write(writer)

    def _writeAutoArrayHolderToStream(self, writer, writeWrongOffset):
        writer.write_bits(self.WRONG_AUTO_ARRAY_OFFSET if writeWrongOffset else self.AUTO_ARRAY_OFFSET, 32)
        writer.write_bits(self.FORCED_ALIGNMENT_VALUE, 8)
        writer.write_varsize(self.AUTO_ARRAY_LENGTH)
        for i in range(self.AUTO_ARRAY_LENGTH):
            writer.write_bits(i, 7)

    def _checkAutoArrayHolder(self, autoArrayHolder, bitPosition=0):
        expectedAutoArrayOffset = (self.AUTO_ARRAY_OFFSET if (bitPosition == 0) else
                                   self.AUTO_ARRAY_OFFSET + (bitPosition // 8))
        self.assertEqual(expectedAutoArrayOffset, autoArrayHolder.auto_array_offset)

        self.assertEqual(self.FORCED_ALIGNMENT_VALUE, autoArrayHolder.force_alignment)

        autoArray = autoArrayHolder.auto_array
        self.assertEqual(self.AUTO_ARRAY_LENGTH, len(autoArray))
        for i in range(self.AUTO_ARRAY_LENGTH):
            self.assertEqual(i, autoArray[i])

    def _createAutoArrayHolder(self, createWrongOffset):
        autoArrayOffset = self.WRONG_AUTO_ARRAY_OFFSET if createWrongOffset else self.AUTO_ARRAY_OFFSET
        autoArray = list(range(self.AUTO_ARRAY_LENGTH))

        return self.api.AutoArrayHolder(autoArrayOffset, self.FORCED_ALIGNMENT_VALUE, autoArray)

    AUTO_ARRAY_LENGTH = 5
    FORCED_ALIGNMENT_VALUE = 0

    WRONG_AUTO_ARRAY_OFFSET = 0
    AUTO_ARRAY_OFFSET = 5

    AUTO_ARRAY_HOLDER_BIT_SIZE = 32 + 1 + 7 + 8 + 5 * 7

File: test/language/offsets/cpp/AutoArrayOffsetTest.cpp
#include "gtest/gtest.h"

#include "offsets/auto_array_offset/AutoArrayHolder.h"

#include "zserio/RebindAlloc.h"
#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

namespace offsets
{
namespace auto_array_offset
{

using allocator_type = AutoArrayHolder::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class AutoArrayOffsetTest : public ::testing::Test
{
protected:
    void writeAutoArrayHolderToByteArray(zserio::BitStreamWriter& writer, bool writeWrongOffset)
    {
        const uint32_t wrongAutoArrayOffset = WRONG_AUTO_ARRAY_OFFSET;
        const uint32_t correctAutoArrayOffset = AUTO_ARRAY_OFFSET;
        const uint32_t autoArrayOffset = (writeWrongOffset) ? wrongAutoArrayOffset : correctAutoArrayOffset;
        writer.writeBits(autoArrayOffset, 32);

        writer.writeBits(FORCED_ALIGNMENT_VALUE, 8);

        writer.writeVarSize(static_cast<uint32_t>(AUTO_ARRAY_LENGTH));
        for (size_t i = 0; i < AUTO_ARRAY_LENGTH; ++i)
            writer.writeBits(static_cast<uint32_t>(i), 7);
    }

    void checkAutoArrayHolder(const AutoArrayHolder& autoArrayHolder, size_t bitPosition = 0)
    {
        const uint32_t expectedAutoArrayOffset = static_cast<uint32_t>(
                (bitPosition == 0) ? AUTO_ARRAY_OFFSET : AUTO_ARRAY_OFFSET + (bitPosition / 8));
        ASSERT_EQ(expectedAutoArrayOffset, autoArrayHolder.getAutoArrayOffset());

        ASSERT_EQ(FORCED_ALIGNMENT_VALUE, autoArrayHolder.getForceAlignment());

        const vector_type<int8_t>& autoArray = autoArrayHolder.getAutoArray();
        ASSERT_EQ(AUTO_ARRAY_LENGTH, autoArray.size());
        for (size_t i = 0; i < AUTO_ARRAY_LENGTH; ++i)
            ASSERT_EQ(i, autoArray[i]);
    }

    void fillAutoArrayHolder(AutoArrayHolder& autoArrayHolder, bool createWrongOffset)
    {
        const uint32_t wrongAutoArrayOffset = WRONG_AUTO_ARRAY_OFFSET;
        const uint32_t correctAutoArrayOffset = AUTO_ARRAY_OFFSET;
        const uint32_t autoArrayOffset = (createWrongOffset) ? wrongAutoArrayOffset : correctAutoArrayOffset;
        autoArrayHolder.setAutoArrayOffset(autoArrayOffset);

        autoArrayHolder.setForceAlignment(FORCED_ALIGNMENT_VALUE);

        vector_type<int8_t> autoArray;
        autoArray.reserve(AUTO_ARRAY_LENGTH);
        for (size_t i = 0; i < AUTO_ARRAY_LENGTH; ++i)
            autoArray.push_back(static_cast<int8_t>(i));
        autoArrayHolder.setAutoArray(autoArray);
    }

    static const size_t     AUTO_ARRAY_LENGTH;
    static const uint8_t    FORCED_ALIGNMENT_VALUE;

    static const uint32_t   WRONG_AUTO_ARRAY_OFFSET;
    static const uint32_t   AUTO_ARRAY_OFFSET;

    static const size_t     AUTO_ARRAY_HOLDER_BIT_SIZE;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const size_t    AutoArrayOffsetTest::AUTO_ARRAY_LENGTH = 5;
const uint8_t   AutoArrayOffsetTest::FORCED_ALIGNMENT_VALUE = 0;

const uint32_t  AutoArrayOffsetTest::WRONG_AUTO_ARRAY_OFFSET = 0;
const uint32_t  AutoArrayOffsetTest::AUTO_ARRAY_OFFSET = 5;

const size_t    AutoArrayOffsetTest::AUTO_ARRAY_HOLDER_BIT_SIZE = 32 + 1 + 7 + 8 + AUTO_ARRAY_LENGTH * 7;

TEST_F(AutoArrayOffsetTest, readConstructor)
{
    const bool writeWrongOffset = false;
    zserio::BitStreamWriter writer(bitBuffer);
    writeAutoArrayHolderToByteArray(writer, writeWrongOffset);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    const AutoArrayHolder autoArrayHolder(reader);
    checkAutoArrayHolder(autoArrayHolder);
}

TEST_F(AutoArrayOffsetTest, readConstructorWrongOffsets)
{
    const bool writeWrongOffset = true;
    zserio::BitStreamWriter writer(bitBuffer);
    writeAutoArrayHolderToByteArray(writer, writeWrongOffset);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    EXPECT_THROW(AutoArrayHolder autoArrayHolder(reader), zserio::CppRuntimeException);
}

TEST_F(AutoArrayOffsetTest, bitSizeOf)
{
    const bool createWrongOffset = false;
    AutoArrayHolder autoArrayHolder;
    fillAutoArrayHolder(autoArrayHolder, createWrongOffset);
    ASSERT_EQ(AUTO_ARRAY_HOLDER_BIT_SIZE, autoArrayHolder.bitSizeOf());
}

TEST_F(AutoArrayOffsetTest, bitSizeOfWithPosition)
{
    const bool createWrongOffset = false;
    AutoArrayHolder autoArrayHolder;
    fillAutoArrayHolder(autoArrayHolder, createWrongOffset);

    const size_t bitPosition = 2;
    ASSERT_EQ(AUTO_ARRAY_HOLDER_BIT_SIZE - bitPosition, autoArrayHolder.bitSizeOf(bitPosition));
}

TEST_F(AutoArrayOffsetTest, initializeOffsets)
{
    const bool createWrongOffset = true;
    AutoArrayHolder autoArrayHolder;
    fillAutoArrayHolder(autoArrayHolder, createWrongOffset);

    const size_t bitPosition = 0;
    ASSERT_EQ(AUTO_ARRAY_HOLDER_BIT_SIZE, autoArrayHolder.initializeOffsets(bitPosition));
    checkAutoArrayHolder(autoArrayHolder);
}

TEST_F(AutoArrayOffsetTest, initializeOffsetsWithPosition)
{
    const bool createWrongOffset = true;
    AutoArrayHolder autoArrayHolder;
    fillAutoArrayHolder(autoArrayHolder, createWrongOffset);

    const size_t bitPosition = 2;
    ASSERT_EQ(AUTO_ARRAY_HOLDER_BIT_SIZE, autoArrayHolder.initializeOffsets(bitPosition));
    checkAutoArrayHolder(autoArrayHolder, bitPosition);
}

TEST_F(AutoArrayOffsetTest, write)
{
    const bool createWrongOffset = false;
    AutoArrayHolder autoArrayHolder;
    fillAutoArrayHolder(autoArrayHolder, createWrongOffset);

    zserio::BitStreamWriter writer(bitBuffer);
    autoArrayHolder.write(writer);
    checkAutoArrayHolder(autoArrayHolder);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    const AutoArrayHolder readAutoArrayHolder(reader);
    checkAutoArrayHolder(readAutoArrayHolder);
    ASSERT_TRUE(autoArrayHolder == readAutoArrayHolder);
}

TEST_F(AutoArrayOffsetTest, writeWithPosition)
{
    const bool createWrongOffset = true;
    AutoArrayHolder autoArrayHolder;
    fillAutoArrayHolder(autoArrayHolder, createWrongOffset);

    const size_t bitPosition = 2;
    zserio::BitStreamWriter writer(bitBuffer);
    writer.writeBits(0, bitPosition);
    autoArrayHolder.initializeOffsets(writer.getBitPosition());
    autoArrayHolder.write(writer);

    checkAutoArrayHolder(autoArrayHolder, bitPosition);
}

TEST_F(AutoArrayOffsetTest, writeWrongOffset)
{
    const bool createWrongOffset = true;
    AutoArrayHolder autoArrayHolder;
    fillAutoArrayHolder(autoArrayHolder, createWrongOffset);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(autoArrayHolder.write(writer), zserio::CppRuntimeException);
}

} // namespace auto_array_offset
} // namespace offsets
